[MIRAI] Annotate panics as either unreachable or as todos

hermanventer opened the pull request:
## Motivation

MIRAI complains about reachable panics. There a good reasons for this because panics are pretty pervasive and are often unexpected and indicate programming mistakes if they are reached.

This PR fixes some calls to panic! to instead be calls to unreachable!, as already suggested in the coding guidelines.

The remaining panics are replaced by a new macro unrecoverable! which just calls panic but prepends the tag "unrecoverable: " to the message. This makes it possible for MIRAI to not warn about these panics and should also make it easier to make sense of panics that show up in logs.

This PR only addressed panics that are currently complained about by MIRAI. There are many more, but I want to keep this PR as small as possible.

There a small number of additional annotations to satisfy MIRAI as well as some comments to help the reader understand why certain panics are in fact unreachable.

I've also removed an unimplemented method. It is not used and keeping it around is both a trap and another cause for a MIRAI message.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

cargo xtest

mimoo commented on types/src/proof/position/mod.rs:
```diff
@@ -89,11 +90,13 @@ impl Position {
 
     /// What is the left node of this node? Will overflow if the node is a leaf
     pub fn left_child(self) -> Self {
+        checked_precondition!(!self.is_leaf());
```
@zamsden ?

hermanventer replied:
You can make this an unchecked precondition if you don't want to pay the runtime cost. This method calls child, which does check the precondition. This annotation adds documentation.

zamsden replied:
It's fine to make this a checked_precondition.  This would be a logic error which should halt execution.

mimoo commented on config/src/trusted_peers.rs:
```diff
@@ -81,6 +81,7 @@ pub struct ConsensusPeersConfig {
 pub struct UpstreamPeersConfig {
     /// List of PeerIds serialized as string.
     pub upstream_peers: Vec<String>,
+    // invariant self.upstream_peers.all(|peer_id_str| { PeerId::from_str(peer_id_str) })
```
did you mean to have that here?

hermanventer replied:
It is documentation that reminds people that users of this field expect all of the strings to be parsable as valid PeerId instances.

mimoo commented:
That's a lot of TODO added :o what's the plan for them?

hermanventer commented:
It would be great if someone (ahem, not me) would create an issue for each of them. They DO need to get fixed. I'll leave the priority of this up to others, but the longer this goes unfixed, the noisier the output from MIRAI will be.

On the whole I'm quite comfortable for CI to run MIRAI and for these messages to get into the faces of developers until it bubbles up in their priority list.

This is not just documentation. It is a fundamentally wrong way of implementing LIBRA and it absolutely must stop NOW.

mimoo commented on config/config-builder/src/swarm_config.rs:
```diff
@@ -175,12 +176,15 @@ impl SwarmConfig {
             warn!("Looking for peer id for peer: {}", node_id);
+            // TODO: Reading the code to figure out why remove_entry below is never going to fail
+            // is tedious and the conclusion that it won't fail is tentative, fragile and subject to change.
+            // Find a better way to be sure that this won't fail.
```
1. if this fails, this would fail directly at start up
2. this is not code running in prod, swarm is only use to test things

I think it's worth removing this TODO.

hermanventer replied:
Testing code needs to be carefully written as well. If it fails in non-obvious ways it can consume a serious amount of developer time.

mimoo commented on consensus/src/chained_bft/block_storage/block_store.rs:
```diff
@@ -184,6 +184,7 @@ impl<T: Payload> BlockStore<T> {
                 )
             })
             .collect();
+        // TODO: if the await returns a failure, propagate it. panic is not an option.
```
@zekun000 what's the alternative here? It sounds like if we can't commit a block that must be committed, we can't make progress either. I'm not sure what else can be done?

hermanventer replied:
There is always something else one can do than panic. If, in fact, this panic will never ever happen then change this to assume_unreachable.

mimoo commented on types/src/access_path.rs:
```diff
@@ -137,6 +139,7 @@ impl Accesses {
 
     /// Return the last access in the sequence
     pub fn last(&self) -> &Access {
+        assume!(self.0.last().is_some()); // follows from invariant
```
that's not always true right? What if I create an access with `Accesses::empty()` and then end up calling this function?

hermanventer replied:
It is kind of a chicken-or-egg situation. Because of this usage, it would be an error to create an empty instance of Accesses, which is what the invariant in line 106 documents. It is also documented in the line below. Currently the code respects this requirement. Currently we can't enforce this, but at some point we could. Meanwhile it helps to warn developers via the comment.

mimoo commented on storage/libradb/src/lib.rs:
```diff
@@ -134,6 +134,8 @@ impl LibraDB {
         let path = db_root_path.as_ref().join("libradb");
         let instant = Instant::now();
+        // TODO: TODO: either return a Result or let the caller open the DB and deal with the failure.
+        // Panic is not an option here.
```
@zamsden  again?

mimoo replied:
My guess is that we only do this at initialization, so it might not be a big deal if we fail here.

mimoo commented on storage/jellyfish-merkle/src/node_type/mod.rs:
```diff
@@ -458,7 +459,8 @@ impl InternalNode {
                             .unwrap_or_else(|| {
-                                panic!(
+                                // Should be guaranteed by the self invariants, but these are not easy to express at the moment
+                                assume_unreachable!(
```
so wouldn't we want MIRAI to actually check these instead of assuming these are unreachable?

hermanventer replied:
assume_unreachable! does a runtime check. It is equivalent to panic a runtime. But at developer time it documents that this should never happen. Panic makes no such statement.

mimoo commented on language/compiler/src/main.rs:
```diff
@@ -72,6 +72,7 @@ fn do_verify_module(module: CompiledModule, dependencies: &[VerifiedModule]) ->
 }
 
 fn write_output(path: &PathBuf, buf: &[u8]) {
+    // TODO: panic is not option here. Provide feedback via println! and fail gracefully.
```
perhaps @sblackshear or somebody else from language should look at these TODOs

hermanventer replied:
If you create an issue and assign it to Sam, that should do the trick.

mimoo commented on consensus/src/chained_bft/persistent_storage.rs:
```diff
@@ -304,6 +304,8 @@ impl<T: Payload> PersistentStorage<T> for StorageWriteProxy {
             .update_to_latest_ledger(0, vec![])
             .expect("unable to read ledger info from storage");
+        // TODO: propagate the failure and let the caller recover or fail gracefully with decent error.
+        // panic is not an option
```
I don't think you can recover from this, you should always have a ledger info in storage (if not the genesis one) IIUC, and so you can't start persistent storage without that info. @wqfish ?

hermanventer replied:
"or fail gracefully with decent error"

mimoo commented:
I CC'ed a few people to look at the different TODOs.

> The appropriate way to deal with expected (but rare) failures is not to panic! and terminate the process. Instead failure values should be propagated and at some point human readable output should be logged and the process terminated gracefully and restarted.

Panicking can be a graceful way to terminate the program, and restart should be figured out by some wrapper not by Libra core itself IMO. So I'm not sure I agree with this.

@bors-libra delegate+

bors-libra commented:
:v: @hermanventer can now approve this pull request

hermanventer commented:
There is room for different opinions on this, but will be extremely helpful to not terminate the program with the panic macro. That macro has a history of abuse and should be flagged as a code smell wherever it occurs. For the cases where the considered opinion is that disgracefully terminating the program with a text message is the best possible way to do things, please invent another macro.

hermanventer commented:
"panic does provide human readable error log and backtrace"

I'm none to sure about this. If this is the case for LIBRA, because of some configuration magic, then fine. Just make sure that by logging you mean something other than spew on the command line.

process::exit(some code with meaning) is somewhat more graceful than panic. It also makes your intent clear.

By "abuse of panic" I mean not making your intent obvious to the reader.

Yes, all I am asking is for there to be coding guidelines that everyone gets to follow and that tooling can enforce.

This PR is meant to start that discussion.

sunshowers commented:
> I'm none to sure about this. If this is the case for LIBRA, because of some configuration magic, then fine. Just make sure that by logging you mean something other than spew on the command line.

FWIW, see https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/controlling-panics-with-std-panic.html.

I do think that for *unexpected* errors which indicate programming bugs, panics are the appropriate way to deal with them.

hermanventer commented:
"I do think that for unexpected errors which indicate programming bugs, panics are the appropriate way to deal with them."

Yes, but by a name other than "panic". In the PR I suggest using "assume_unreachable", which is understood by MIRAI to be a considered statement by the developer that reasoning beyond what the MIRAI is capable of leads to the conclusion that this should never happen.

Good practice is to document this reasoning in a comment.

sunshowers commented:
Got it, thank you for clarifying!

aching commented:
> Yes, all I am asking is for there to be coding guidelines that everyone gets to follow and that tooling can enforce.

@hermanventer, can you please make changes to https://github.com/libra/libra/blob/master/documentation/coding_guidelines.md#error-handling as per your suggestions?  I think it would help with discussion on this PR.

hermanventer commented:
I hope that I have now addressed all of the feedback in satisfactory way. Please have another look. In particular, see if you agree that panics that are marked as unreachable are in fact unreachable.

metajack commented:
@mimoo @aching Is someone going to re-review this with the new change?

mimoo commented:
@zekun000  are the changes acceptable?

zamsden commented on storage/jellyfish-merkle/src/node_type/mod.rs:
```diff
@@ -380,7 +380,7 @@ impl InternalNode {
             self.child(only_child_index)
                 .unwrap_or_else(|| {
-                    panic!(
+                    unreachable!(
```
This is not unreachable.  This is a legitimately reachable state from which there is no recovery - we have a corrupt tree on storage and cannot continue.

A panic is the best possible option, IMO.  Edit maybe unrecoverable! is the better choice.

zamsden commented on storage/jellyfish-merkle/src/node_type/mod.rs:
```diff
@@ -458,7 +458,8 @@ impl InternalNode {
                             .unwrap_or_else(|| {
-                                panic!(
+                                // Should be guaranteed by the self invariants, but these are not easy to express at the moment
+                                unreachable!(
```
Same thing, this is also not unreachable, this is also unrecoverable.

zamsden commented on storage/jellyfish-merkle/src/tree_cache/mod.rs:
```diff
@@ -213,7 +213,7 @@ where
         let root_hash = self
             .get_node(root_node_key)
-            .unwrap_or_else(|_| panic!("Root node with key {:?} must exist", root_node_key))
+            .unwrap_or_else(|_| unreachable!("Root node with key {:?} must exist", root_node_key))
```
This should be genuinely unreachable, but I would need a little more scrutiny.

hermanventer replied:
By all means. An annotation such as unreachable! should always be an invitation to double check. After all, a human is making an assertion that something cannot actually happen. Sometimes humans are wrong.

zamsden commented on types/src/proof/position/mod.rs:
```diff
@@ -89,11 +90,13 @@ impl Position {
 
     /// What is the right node of this node? Will overflow if the node is a leaf
     pub fn right_child(self) -> Self {
+        checked_precondition!(!self.is_leaf());
```
Same thing.

zekun000 commented:
consensus related looks good, but requiring every team to stamp is hard. 
usually I'd break the big refactor diff based on folders that each team owns so that can move faster

mimoo commented on common/failure-ext/failure-macros/src/lib.rs:
```diff
@@ -13,3 +13,13 @@ macro_rules! bail_err {
+/// Use this for errors that arise in correct programs due to external factors.
+/// For example, if a file that is essential for running cannot be found for some reason.
+#[macro_export]
+macro_rules! unrecoverable {
```
this is a great idea btw!

mimoo commented on documentation/coding_guidelines.md:
```diff
@@ -186,7 +186,7 @@ Error handling suggestions follow the [Rust book guidance](https://doc.rust-lang
 *Panic*
 
-* `panic!()` - Runtime panic! should only be used when the resulting state cannot be processed going forward.  It should not be used for any recoverable errors.
+* `unrecoverable!()` - Logs an error! causes a panic! Should only be used when the resulting state cannot be processed going forward.
```
I'm in support of this change, but to have even more impact I think we should lint panics and point developers to use unrecoverable instead. I can try to do this, would x lint be a good place to do this? @bmwill

mimoo commented:
It looks good to me, so I think we shouldn't spent more time on this. Please:

* take a look at what @zamsden pointed out.
* let's think how we can push for unrecoverable instead, maybe by linting panics (that's on me unless there is any takers).
* think about the assumes, and how that could bite us in the face later if we are too aggressive with them :)

@bors-libra delegate+

bors-libra commented:
:v: @hermanventer can now approve this pull request

bors-libra commented:
:v: @hermanventer can now approve this pull request

bors-libra commented:
:v: @hermanventer can now approve this pull request

hermanventer commented:
@bors-libra r=mimoo

bors-libra commented:
:pushpin: Commit ae93d5d has been approved by `mimoo`

<!-- @bors-libra r=mimoo ae93d5d86550b03f6a59db6bc23ad5328b56a881 -->

bors-libra commented:
:hourglass: Testing commit ae93d5d86550b03f6a59db6bc23ad5328b56a881 with merge 163b94420bcff8bbfcdc5fc6d9a25e3fc9360f69...

bors-libra commented:
:sunny: Test successful - [checks-circle_commit_workflow](https://circleci.com/workflow-run/cf30da44-a041-49a1-8ed9-fc187618a887?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: mimoo
Pushing 163b94420bcff8bbfcdc5fc6d9a25e3fc9360f69 to master...

bmwill commented:
@hermanventer Since this new macro exists to please mirai, maybe it would make sense to move the macro to the mirai-annotations crate instead?

hermanventer commented:
> @hermanventer Since this new macro exists to please mirai, maybe it would make sense to move the macro to the mirai-annotations crate instead?

It makes sense to add it to the annotations crate and I'll do so. Meanwhile adding it as in this PR was by far the least intrusive change.

I also want to emphasize that pleasing MIRAI is not an end in itself. I contend that the changes in this PR makes the code more readable and that this is good in itself.
